### PR TITLE
Fix fieldSpec when exporting cases using certain (default) fields

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
`case_activity_details` is a default field when using Find Cases->Export (primary fields).

But `case_activity_details` is not a real database field. It is really `civicrm_activity.activity_details`. The error is normally hidden because the default field type for export is "text" if there is no fieldSpec. However, `activity_details` is longtext so if you have a value in the DB that is longer than "text" supports the export will fail with a DB error:

`Data too long for column 'case_activity_details' at..`

Before
----------------------------------------
Case exports may fail because export uses the wrong datatype for temporary table column.

After
----------------------------------------
Case exports won't fail because of wrong datatype in temporary table column.

Technical Details
----------------------------------------
I've added a `getFieldAliases()` function to map them. There must be a better way to do this, at a higher level via metadata?

Comments
----------------------------------------
@eileenmcnaughton @demeritcowboy

### CODE REVIEW

**Comment by demeritcowboy** [comment]
I don't think I have much to contribute here, other than that yes probably TEXT is too small. As a general note, for export I'm not sure how useful `details` is in practice, especially since the field is often wordy and not very rigorous and is filled with html like `<p>` . Should it even be part of "exportableFields"?

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Can confirm the problem and it exports my giant field.
screenshot

* Developer standards
    * [`r-tech`] __?????__
      * Not sure what to say. I've never really looked at the exportprocessor. This export is really an activity export with some joined case data, and I guess I agree with that concept, but it sounds like it's done the other way around?
    * [`r-code`] __PASS__
    * [`r-maint`] __Undecided__
    * [`r-test`] __PASS__

**Comment by eileenmcnaughton** [comment]
So wrt the exportprocessor - in general the way we have been working is to try to get the fields correctly defined on the BAO_Query class & use them on the processor. In this case I think they need to all be defined correctly  in in BAO_Activity::exportableFields and there shouldn't need to be any hacks directly on the ExportProcessor

e.g pseudocode

```
diff --git a/CRM/Activity/BAO/Activity.php b/CRM/Activity/BAO/Activity.php
index 20d2f0e072..24e4f48917 100644
--- a/CRM/Activity/BAO/Activity.php
+++ b/CRM/Activity/BAO/Activity.php
@@ -2168,49 +2168,10 @@ AND cl.modified_id  = c.id
       $fields['activity_type_id']['title'] = ts('Activity Type ID');
     }
     else {
-      // Set title to activity fields.
-      $fields = [
-        'case_activity_subject' => [
-          'title' => ts('Activity Subject'),
-          'type' => CRM_Utils_Type::T_STRING,
-        ],
-        'case_source_contact_id' => [
-          'title' => ts('Activity Reporter'),
-          'type' => CRM_Utils_Type::T_STRING,
-        ],
-        'case_recent_activity_date' => [
-          'title' => ts('Activity Actual Date'),
-          'type' => CRM_Utils_Type::T_DATE,
-        ],
-        'case_scheduled_activity_date' => [
-          'title' => ts('Activity Scheduled Date'),
-          'type' => CRM_Utils_Type::T_DATE,
-        ],
-        'case_recent_activity_type' => [
-          'title' => ts('Activity Type'),
-          'type' => CRM_Utils_Type::T_STRING,
-        ],
-        'case_activity_status' => [
-          'title' => ts('Activity Status'),
-          'type' => CRM_Utils_Type::T_STRING,
-        ],
-        'case_activity_duration' => [
-          'title' => ts('Activity Duration'),
-          'type' => CRM_Utils_Type::T_INT,
-        ],
-        'case_activity_medium_id' => [
-          'title' => ts('Activity Medium'),
-          'type' => CRM_Utils_Type::T_INT,
-        ],
-        'case_activity_details' => [
-          'title' => ts('Activity Details'),
-          'type' => CRM_Utils_Type::T_TEXT,
-        ],
-        'case_activity_is_auto' => [
-          'title' => ts('Activity Auto-generated?'),
-          'type' => CRM_Utils_Type::T_BOOLEAN,
-        ],
-      ];
+      foreach ('subject', 'source_contact_id', ... as $realName) {
+        $fields['case' . $realName] => $fields[$realName];
+      }
+

```

**Comment by mattwire** [comment]
Thanks @eileenmcnaughton @demeritcowboy I've updated this so it should be better now?

**Comment by eileenmcnaughton** [comment]
I'm OK with this if it passes tests & @demeritcowboy confirms it still works.

I think test cover should be good & this isn't adding complexity which is great

**Comment by demeritcowboy** [comment]
jenkins retest this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@demeritcowboy @mattwire test fails are related - this might work - https://github.com/civicrm/civicrm-core/pull/17537

I haven't test but am working on the general theory that you don't have to limit the fields defined on the query object as it is the available fields, not necessarily the used ones

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by demeritcowboy** [comment]
Those priority_id test fails are the same ones I saw before. Since it's a consistent fail it must be real. But I can't see anything obvious why it affects priority_id.

**Comment by mattwire** [comment]
Test failure is now test-1 mysql not working..

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
CRM_Export_BAO_ExportTest::testGetSQLColumnsAndHeaders with data set #2 (6, array('`civicrm_primary_id` varchar(16)', '`contact_type` varchar(64)', '`contact_sub_type` varchar(255)', '`do_not_email` varchar(16)', '`do_not_phone` varchar(16)', '`do_not_mail` varchar(16)', '`do_not_sms` varchar(16)', '`do_not_trade` varchar(16)', '`is_opt_out` varchar(16)', '`legal_identifier` varchar(32)', '`external_identifier` varchar(64)', '`sort_name` varchar(128)', '`display_name` varchar(128)', '`nick_name` varchar(128)', '`legal_name` varchar(128)', '`image_url` longtext', '`preferred_communi

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -516,9 +516,10 @@ public function __construct(
       $this->_fields = array_merge($this->_fields, $fields);
 
       // add activity fields
-      $this->_fields = array_merge($this->_fields, CRM_Activity_BAO_Activity::exportableFields());
-      // Add hack as no unique name is defined for the field but the search form is in denial.
-      $this->_fields['activity_priority_id'] = $this->_fields['priority_id'];
+      $this->_fields = array_merge(
+        $this->_fields,
+        CRM_Activity_BAO_Activity::exportableFields(CRM_Contact_Form_Search::getModeValue($this->_mode)['entity'])
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php:521]: I think given we only support 'Activity' and 'Case' we probably should stick with hard-coding. I like more metadata but it also feels like it's opening up flakey code to have implied support for more than just the 2 things it does support & there is no particular benefit it NOT defining case activity fields when it isn's a case activity search

**Comment by eileenmcnaughton** [comment]
This has been stalled on failing tests for a while - I put up an alternate that addresses them https://github.com/civicrm/civicrm-core/pull/17970

**Comment by eileenmcnaughton** [comment]
Closing in favour of  #17970 since I got the tests passing on that one

**Comment by mattwire** [comment]
Thanks @eileenmcnaughton @demeritcowboy

**Pull request closed without merging**